components/brave_federated/learning_service.cc

```diff
@@ -0,0 +1,173 @@
+#include "brave/components/brave_federated/task/typing.h"
+#include "brave/components/brave_federated/util/synthetic_dataset.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+
+namespace brave_federated {
+
+LearningService::LearningService(
+    EligibilityService* eligibility_service,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory),
+      eligibility_service_(eligibility_service) {
+  DCHECK(url_loader_factory);
+  DCHECK(eligibility_service);
+
+  communication_adapter_ = new CommunicationAdapter(url_loader_factory);
+  eligibility_service_->AddObserver(this);
+
+  StartParticipating();
+}
+
+LearningService::~LearningService() {
+  StopParticipating();
+  eligibility_service_->RemoveObserver(this);
+}
+
+void LearningService::StartParticipating() {
+  participating_ = true;
+  GetTasks();
+}
+
+void LearningService::StopParticipating() {
+  DCHECK(participating_);
+  participating_ = false;
+}
+
+void LearningService::GetTasks() {
+  communication_adapter_->GetTasks(base::BindOnce(
+      &LearningService::HandleTasksOrReconnect, base::Unretained(this)));
+}
+
+void LearningService::HandleTasksOrReconnect(TaskList tasks, int reconnect) {
+  if (tasks.size() == 0) {
+    reconnect_timer_ = std::make_unique<base::RetainingOneShotTimer>();
+    reconnect_timer_->Start(FROM_HERE, base::Seconds(reconnect), this,
+                            &LearningService::GetTasks);
+    VLOG(2) << "No tasks available, reconnecting in " << reconnect << "s";
+    return;
+  }
+
+  // TODO(lminto): for now, disregard all tasks beyond the first one
+  // TODO(stevelaskaridis): extract into a utility function
+  Task task = tasks.at(0);
+  float lr = 0.01;
+  std::map<std::string, float> config = task.GetConfig();
+  auto cursor = config.find("lr");
+  if (cursor != config.end()) {
+    lr = cursor->second;
+    VLOG(2) << "Learning rate applied from server: " << lr;
+  }
+
+  ModelSpec spec{
+      32,    // num_params
+      64,    // batch_size
+      lr,    // learning_rate
+      500,   // num_iterations
+      0.5    // threshold
+  };
+
+  if (spec.num_params != static_cast<int>(task.GetParameters().at(0).size())) {
+    VLOG(2) << "Task specifies a different model size than the client";
+    return;
+  }
+  VLOG(2) << "Task model and client model match!";
+
+  Model* model = new Model(spec);
+  model->SetWeights(task.GetParameters().at(0));
+  model->SetBias(task.GetParameters().at(1).at(0));
+  FederatedTaskRunner* task_runner = new FederatedTaskRunner(task, model);
+
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE,
+      {base::TaskPriority::BEST_EFFORT,
+       base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN},
+      base::BindOnce(
+          [](FederatedTaskRunner* task_runner) -> SyntheticDataset* {
```

Comment:
In fact the created `SyntheticDataset` is never deleted. unique_ptr should be used here